ui/models.py

```diff
@@ -42,45 +44,104 @@ class Video(models.Model):
     description = models.TextField(blank=True)
     source_url = models.URLField()
     moira_lists = models.ManyToManyField(MoiraList)
+    status = models.TextField(max_length=24, blank=True)
+    s3_subkey = models.UUIDField(unique=True, null=False, blank=False, default=uuid4)
+
+    @property
+    def s3_key(self):
+        """
+        Avoid duplicate S3 keys/filenames when transferring videos rom Dropbox
+
+        Args:
+            user_id (str): The user who is uploading a new video
+            key (str): The base key to check for duplicates
+
+        Returns:
+            str: A unique S3 key including the user id as a virtual subfolder
+        """
+        basename, extension = os.path.splitext(self.source_url.split('/')[-1])
+        newkey = '{user}/{uuid}/{base}{ext}'.format(
+            user=self.creator.id, uuid=str(self.s3_subkey), base=basename, ext=extension)
+        return newkey
+
+
+class VideoFile(models.Model):
+    """
+    A file associated with a Video object, either the original upload or a transcoded file.
+    """
+    created_at = models.DateTimeField(auto_now_add=True)
+    s3_object_key = models.TextField(unique=True)
+    bucket_name = models.CharField(max_length=128, blank=False, null=False)
+    video = models.ForeignKey(Video, on_delete=models.CASCADE)
+    encoding = models.CharField(max_length=128, default='original')
+    preset_id = models.CharField(max_length=128, blank=True, null=True)
 
     @property
     def s3_object(self):
-        """s3 object"""
-        bucket_name = os.environ.get("VIDEO_S3_BUCKET", "odl-video-service")
+        """
+        Get the S3 Object for the video
+
+            Returns:
+                s3.Object: Video file's S3 object
+        """
         s3 = boto3.resource('s3')
-        return s3.Object(bucket_name, self.s3_object_key)
+        return s3.Object(self.bucket_name, self.s3_object_key)
```

Comment:
I think by ensuring that bucket_name and object_key database fields are never null, there shouldn't be any exception here to handle, only later when this is used to access the object in some way, for example ```videofile.s3_object.delete()```